feature/amazon/src/test/java/com/revenuecat/purchases/amazon/handler/ProductDataHandlerTest.kt

```diff
@@ -0,0 +1,270 @@
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class ProductDataHandlerTest {
+
+    private lateinit var underTest: ProductDataHandler
+    private val apiKey = "api_key"
+    private lateinit var cache: MockDeviceCache
+    private lateinit var purchasingServiceProvider: PurchasingServiceProviderForTest
+
+    @Before
+    fun setup() {
+        cache = MockDeviceCache(mockk(), apiKey)
+        purchasingServiceProvider = PurchasingServiceProviderForTest()
+        underTest = ProductDataHandler(purchasingServiceProvider)
+    }
+
+    @Test
+    fun `send cached product data if available for all skus`() {
+        val expectedSkus = setOf("sku_a", "sku_b")
+
+        val expectedProductDataInCache = mapOf(
+            "sku_a" to dummyAmazonProduct(sku = "sku_a"),
+            "sku_b" to dummyAmazonProduct(sku = "sku_b")
+        )
+
+        underTest.productDataCache.putAll(expectedProductDataInCache)
+
+        var receivedProductList: List<ProductDetails>? = null
+        underTest.getProductData(
+            expectedSkus,
+            "US",
+            onReceive = {
+                receivedProductList = it
+            },
+            onError = {
+                fail("should be success")
+            }
+        )
+
+        assertThat(receivedProductList).isNotNull
+        expectedSkus.forEach { expectedSku ->
+            assertThat(receivedProductList!!.firstOrNull { expectedSku == it.sku }).isNotNull
+        }
+
+        assertThat(purchasingServiceProvider.getProductDataCalled).isFalse
+    }
+
+    @Test
+    fun `getProductData is called if one of the products is not in cache`() {
+        val firstProduct = "sku_a" to dummyAmazonProduct(sku = "sku_a")
+        val expectedProductData = mapOf(
+            firstProduct,
+            "sku_b" to dummyAmazonProduct(sku = "sku_b")
+        )
+
+        underTest.productDataCache.putAll(mapOf(firstProduct))
+
+        val dummyRequestId = "a_request_id"
+        purchasingServiceProvider.getProductDataRequestId = dummyRequestId
+
+        var receivedProductList: List<ProductDetails>? = null
+
+        underTest.getProductData(
+            expectedProductData.keys,
+            "US",
+            onReceive = {
+                receivedProductList = it
+            },
+            onError = {
+                fail("should be success")
+            }
+        )
+
+        underTest.onProductDataResponse(
+            getDummyProductDataResponse(
+                dummyRequestId,
+                productData = expectedProductData
+            )
+        )
+
+        assertThat(receivedProductList).isNotNull
+        expectedProductData.keys.forEach { expectedSku ->
+            assertThat(receivedProductList!!.firstOrNull { expectedSku == it.sku }).isNotNull
+        }
+
+        assertThat(purchasingServiceProvider.getProductDataCalled).isTrue
+    }
+
+    @Test
+    fun `products are cached after a successful response`() {
```

Comment:
should we also have a test to ensure that the cache works? i.e.: that if it gets called twice, the second time it hits cache?